# Завдання 1.**Текст, выделенный полужирным шрифтом**

In [18]:
class HashTable:
    """
    Реалізація хеш-таблиці in Python.

    Атрибути:
        size: Розмір хеш-таблиці.
        table: Список, що представляє хеш-таблицю. Кожен елемент списку - це список пар (ключ, значення) для певного індексу хешування.
    """

    def __init__(self, size):
        """
        Ініціалізація хеш-таблиці.

        Args:
            size: Розмір хеш-таблиці.
        """
        self.size = size
        self.table = [[] for _ in range(self.size)]

    def hash_function(self, key):
        """
        Функція хешування для ключів.

        Args:
            key: Ключ для хешування.

        Returns:
            Індекс у хеш-таблиці, отриманий за допомогою вбудованої функції hash(key) та операції mod розміром хеш-таблиці.
        """
        return hash(key) % self.size

    def insert(self, key, value):
        """
        Вставка пари (ключ, значення) в хеш-таблицю.

        Args:
            key: Ключ для пари.
            value: Значення для пари.

        Returns:
            True - якщо пара успішно вставлена, False - якщо ключ вже існує.
        """
        key_hash = self.hash_function(key)
        key_value = [key, value]

        if not self.table[key_hash]:
            self.table[key_hash] = [key_value]
            return True
        else:
            for pair in self.table[key_hash]:
                if pair[0] == key:
                    pair[1] = value  # Оновити значення, якщо ключ вже існує
                    return True
            self.table[key_hash].append(key_value)
            return True

    def get(self, key):
        """
        Отримання значення за ключем.

        Args:
            key: Ключ для пошуку.

        Returns:
            Значення пари (ключ, значення), якщо ключ знайдено, None - якщо ключ не знайдено.
        """
        key_hash = self.hash_function(key)
        if self.table[key_hash] is not None:
            for pair in self.table[key_hash]:
                if pair[0] == key:
                    return pair[1]
        return None

    def delete(self, key):
        """
        Видалення пари (ключ, значення) з хеш-таблиці.

        Args:
            key: Ключ для видалення.

        Returns:
            True - якщо пару успішно видалено, False - якщо ключ не знайдено.
        """
        key_hash = self.hash_function(key)
        if self.table[key_hash] is not None:
            for i in range(len(self.table[key_hash])):
                if self.table[key_hash][i][0] == key:
                    del self.table[key_hash][i]
                    return True
        return False

# Тест хеш-таблиці

H = HashTable(5)
H.insert("книга", 100)
H.insert("телефон", 500)
H.insert("комп'ютер", 1500)

print(H.get("книга"))  # Виводить: 100
print(H.get("телефон"))  # Виводить: 500
print(H.get("комп'ютер"))  # Виводить: 1500

H.delete("комп'ютер")
print(H.get("комп'ютер"))  # Виводить: None

H.insert("стіл", 200)
print(H.get("стіл"))  # Виводить: 200


100
500
1500
None
200


# Завдання **2**

In [19]:
def binary_search(arr, target):

    if not arr:
        raise ValueError("Масив не може бути порожнім")

    left, right = 0, len(arr) - 1
    iterations = 0
    upper_bound = None

    while left <= right:
        iterations += 1
        mid = left + (right - left) // 2

        if arr[mid] < target:
            left = mid + 1
        elif arr[mid] > target:
            right = mid - 1
        else:
            # Точно знаходимо target, знаходимо upper_bound
            upper_bound = arr[mid]
            while mid + 1 < len(arr) and arr[mid + 1] == target:
                mid += 1
            if mid + 1 < len(arr):
                upper_bound = arr[mid + 1]
            return iterations, upper_bound

    # Якщо ми не знайшли точний елемент
    if left < len(arr):
        upper_bound = arr[left]

    return iterations, upper_bound

# Тестування
arr = [3.6, 3.0, 0.5, 10.0, 1.7, 1.2]
target = 1.2
print(binary_search(arr, target))  # Приклад, де елемент знайдено

target = 0.1
print(binary_search(arr, target))  # Приклад, де елемент не знайдено, але виведемо верхню межу





(3, 10.0)
(2, 3.6)


# **Завдання 3.**

In [20]:
import timeit

# Реалізація алгоритму Боєра-Мура
def boyer_moore_search(text, pattern):
    m = len(pattern)
    n = len(text)
    if m == 0:
        return 0
    char_table = make_char_table(pattern)
    offset_table = make_offset_table(pattern)
    i = m - 1
    while i < n:
        j = m - 1
        while pattern[j] == text[i]:
            if j == 0:
                return i
            i -= 1
            j -= 1
        i += max(offset_table[m - 1 - j], char_table.get(text[i], m))
    return -1

def make_char_table(pattern):
    table = {}
    for i in range(len(pattern) - 1):
        table[pattern[i]] = len(pattern) - 1 - i
    return table

def make_offset_table(pattern):
    table = []
    last_prefix_position = len(pattern)
    for i in range(len(pattern) - 1, -1, -1):
        if is_prefix(pattern, i + 1):
            last_prefix_position = i + 1
        table.append(last_prefix_position - i + len(pattern) - 1)
    for i in range(len(pattern) - 1):
        slen = suffix_length(pattern, i)
        table[slen] = len(pattern) - 1 - i + slen
    return table

def is_prefix(pattern, p):
    j = 0
    for i in range(p, len(pattern)):
        if pattern[i] != pattern[j]:
            return False
        j += 1
    return True

def suffix_length(pattern, p):
    length = 0
    j = len(pattern) - 1
    i = p
    while i >= 0 and pattern[i] == pattern[j]:
        length += 1
        i -= 1
        j -= 1
    return length

# Реалізація алгоритму Кнута-Морріса-Пратта
def knuth_morris_pratt(text, pattern):
    def compute_lps(pattern):
        lps = [0] * len(pattern)
        length = 0
        i = 1
        while i < len(pattern):
            if pattern[i] == pattern[length]:
                length += 1
                lps[i] = length
                i += 1
            else:
                if length != 0:
                    length = lps[length - 1]
                else:
                    lps[i] = 0
                    i += 1
        return lps

    lps = compute_lps(pattern)
    i = 0
    j = 0
    while i < len(text):
        if pattern[j] == text[i]:
            i += 1
            j += 1
        if j == len(pattern):
            return i - j
        elif i < len(text) and pattern[j] != text[i]:
            if j != 0:
                j = lps[j - 1]
            else:
                i += 1
    return -1

# Реалізація алгоритму Рабіна-Карпа
def rabin_karp(text, pattern, d=256, q=101):
    M = len(pattern)
    N = len(text)
    i = 0
    j = 0
    p = 0  # hash value for pattern
    t = 0  # hash value for text
    h = 1

    for i in range(M-1):
        h = (h * d) % q

    for i in range(M):
        p = (d * p + ord(pattern[i])) % q
        t = (d * t + ord(text[i])) % q

    for i in range(N - M + 1):
        if p == t:
            for j in range(M):
                if text[i + j] != pattern[j]:
                    break
            else:
                return i
        if i < N - M:
            t = (d * (t - ord(text[i]) * h) + ord(text[i + M])) % q
            if t < 0:
                t = t + q
    return -1




In [21]:
# Визначимо текст статті
text1 = """
ВИКОРИСТАННЯ АЛГОРИТМІВ У БІБЛІОТЕКАХ МОВ ПРОГРАМУВАННЯ
Автори публiкації: Коваленко О.О., Корягіна Д.О. Вінницький національний технічний університет
Анотація: У данній статті було розглянуто використання алгоритмів у бібліотеках мов програмування. Ключові слова: алгоритми, сортування, лінійний пошук, двійковий пошук, пошук стрибками, інтерполяційний пошук, експоненціальний пошук, жадібний алгоритм.
Стаття
За всю історію комп'ютерних наук склалося розуміння, які алгоритми та структури даних (способи їх зберігання) потрібні для вирішення практичних завдань, певний набір, який повинен знати кожен розробник. Наприклад, сортування: товари в магазині сортують за вартістю або терміну придатності, а ресторани – за віддаленістю або рейтингу. Хеш-таблиці допомагають перевірити коректність пароля та не зберігати його на сайті у відкритому вигляді, графи – знаходити найкоротший шлях і зберігати зв'язки між користувачами в соцмережах.
Алгоритми – це послідовність точно визначених дій, які призводять до вирішення поставленої задачі чи певного завдання і на сьогодні уже створено величезну кількість алгоритмів для вирішення важких задач, що полегшують написання коду будь-якому програмісту, особливо початківцям [1].
Метою роботи є виявлення найбільш популярних алгоритмів у бібліотеках мов програмування.
Усі алгоритми та структури даних вже давно реалізовані в бібліотеках популярних мов програмування. Більше ніхто не пише вручну алгоритм сортування чисел, а щоб користуватися хеш-таблицями, навіть не потрібно знати, як вони влаштовані.
Але наявність безлічі готових бібліотек не означає, що не потрібно розуміти, як влаштовані алгоритми. Фундаментальні знання допомагають дізнатися, що всередині, як воно працює і чому одне рішення краще, аніж інше у конкретній ситуації. Якщо зрозуміти, як влаштовані класичні алгоритми, то можна створювати власні рішення, комбінувати методи один з одним, щоб вирішувати більш складні завдання.
У програмуванні стандартна бібліотека — це бібліотека, що доступна в усіх реалізаціях даної мови програмування. Зміст такої бібліотеки зазвичай описано у специфікації мови, однак також він може частково або повністю визначатися більш неформальними практиками програмістів, що користуються нею. Більшість стандартних бібліотек включають у себе визначення принаймні таких найчастіше використовуваних інструментів як:
Алгоритми (такі як алгоритми сортування);
Структури даних (наприклад, списки, дерева, хеш-таблиці); ої задачі чи певного завдання і на сьогодні уже створено величезну кількість алгоритмів для вирішення важких задач, що полегшують написання коду будь-якому програмісту, особливо початківцям.
Метою роботи є виявлення найбільш популярних алгоритмів у бібліотеках мов програмування.
"""

text2 = """
Методи та структури даних для реалізації бази даних рекомендаційної системи соціальної мережі
Автори публiкації: Міхав В.В., Мелешко Є.В., Шимко С.В. Центральноукраїнський національний технічний університет,
Анотація
Метою даної роботи є дослідження та програмна реалізація методів і структур даних для побудови бази даних рекомендаційної системи, щоб порівняти ефективність їх використання за затратами часу та пам’яті. Наявність великої кількості різних методів реалізації баз даних викликає необхідність порівняльного аналізу та вибору оптимального методу і структури даних для зберігання інформації у рекомендаційних системах.
Було проведено дослідження різних структур даних, які можна використати для створення бази даних рекомендаційної системи, зокрема, досліджені зв’язний список, розгорнутий зв’язний список, хеш-таблиця, B-дерево, B+-дерево та бінарна діаграма рішень. Також було проведено серію експериментів на програмній імітаційній моделі рекомендаційної системи з різною кількістю агентів, предметів та сесій.
Відповідно до результатів проведених експериментів, розгорнутий список показав найкращі показники швидкодії та використання пам’яті. Структура B+-дерево показала результати, близькі до хеш-таблиці. Час доступу до окремого елементу в обох випадках сталий, але B+- дерево має певні переваги – елементи зберігаються відсортованими, а при зміні розміру немає необхідності розширювати область пам’яті. Найгірші результати показала структура даних бінарна діаграма рішень як за затратами часу, так і за затратами пам’яті. Профілювання показало, що 75% часу роботи тесту варіанту з розгорнутим списком зайняло генерування випадкових даних для програмного імітаційного моделювання агентів та предметів рекомендаційної системи, тож, саме сховище даних має високі показники ефективності.
Профілювання варіанту із інвертованим списком показало, що доступ до випадкових блоків займає більше часу через неможливість закешувати їх, тож, за умов реального навантаження час вставки нових даних буде більшим, а відносна ефективність застосування інвертованого списку зросте. Для найбільш ефективного використання пам’яті розмір блоку зв’язного списку має бути адаптований таким чином, щоб блоки були максимально заповнені. Блоки малого розміру зменшують втрати пам’яті, але збільшують час обходу всіх елементів списку та збільшують накладні витрати пам’яті.
Ключові слова: рекомендаційні системи, бази даних, структури даних, програмна імітаційна модель.
"""

# Вибір підрядків для пошуку
existing_substring = "алгоритм"
fake_substring = "мирне небо"

# Функція для вимірювання часу виконання
def measure_time(func, text, pattern):
    return timeit.timeit(lambda: func(text, pattern), number=10)

# Вимірюємо час для кожного алгоритму
results = {}
for text, label in [(text1, "Text1"), (text2, "Text2")]:
    results[label] = {}
    for algorithm, func in [("Boyer-Moore", boyer_moore_search),
                            ("Knuth-Morris-Pratt", knuth_morris_pratt),
                            ("Rabin-Karp", rabin_karp)]:
        results[label][algorithm] = {
            "existing": measure_time(func, text, existing_substring),
            "fake": measure_time(func, text, fake_substring)
        }

# Виводимо результати
results

{'Text1': {'Boyer-Moore': {'existing': 0.0005077089999758755,
   'fake': 0.003527180000219232},
  'Knuth-Morris-Pratt': {'existing': 0.0022455189996435365,
   'fake': 0.017616926000300737},
  'Rabin-Karp': {'existing': 0.0011127319999104657,
   'fake': 0.014581930000076682}},
 'Text2': {'Boyer-Moore': {'existing': 0.002220697999746335,
   'fake': 0.0019461229999251373},
  'Knuth-Morris-Pratt': {'existing': 0.01414991799993004,
   'fake': 0.014281587999903422},
  'Rabin-Karp': {'existing': 0.01605844400000933,
   'fake': 0.01332305800042377}}}


На основі наведених даних про ефективність алгоритмів пошуку підрядка Боєра-Мура, Кнута-Морріса-Пратта та Рабіна-Карпа в двох текстових файлах (Text1 і Text2), ми можемо зробити наступні висновки:

Text1
Для тексту Text1 було виміряно час виконання для кожного з алгоритмів на підрядках, що існують в тексті та вигаданих підрядках:

Боєра-Мура:
Існуючий: 0.000507709 с
Вигаданий: 0.003527180 с
Кнута-Морріса-Пратта:
Існуючий: 0.002245519 с
Вигаданий: 0.017616926 с
Рабіна-Карпа:
Існуючий: 0.001112732 с
Вигаданий: 0.014581930 с

Text2
Для тексту Text2 було виміряно час виконання для кожного з алгоритмів на підрядках, що існують в тексті та вигаданих підрядках:

Боєра-Мура:
Існуючий: 0.002220698 с
Вигаданий: 0.001946123 с
Кнута-Морріса-Пратта:
Існуючий: 0.014149918 с
Вигаданий: 0.014281588 с
Рабіна-Карпа:
Існуючий: 0.016058444 с
Вигаданий: 0.013323058 с

Висновки

Text1:

Найшвидший алгоритм для існуючого підрядка: Боєра-Мура (0.000507709 с)
Найшвидший алгоритм для вигаданого підрядка: Боєра-Мура (0.003527180 с)
В загальному, алгоритм Боєра-Мура виявився найшвидшим як для існуючого, так і для вигаданого підрядка.
Text2:

Найшвидший алгоритм для існуючого підрядка: Боєра-Мура (0.002220698 с)
Найшвидший алгоритм для вигаданого підрядка: Боєра-Мура (0.001946123 с)
В загальному, алгоритм Боєра-Мура також виявився найшвидшим для обох видів підрядків.

Загальний висновок:

Алгоритм Боєра-Мура показав найкращі результати як для тексту Text1, так і для тексту Text2 для обох типів підрядків (існуючого та вигаданого). Це свідчить про його високу ефективність у порівнянні з алгоритмами Кнута-Морріса-Пратта та Рабіна-Карпа в даних умовах.